Dockerfile

```diff
@@ -0,0 +1,28 @@
+FROM ubuntu:latest
+
+# Install environment dependencies
+RUN apt update && apt install -y wget unzip python3 python3-pip python3-dev python-is-python3 default-jdk nodejs npm
+RUN python3 -m pip install -U pip
+
+# Setup build environment
+ENV BAZEL_VERSION='4.0.0'
+ENV BAZEL_SHA256SUM='7bee349a626281fc8b8d04a7a0b0358492712377400ab12533aeb39c2eb2b901'
+ENV BUILDTOOLS_VERSION='3.0.0'
+ENV BUILDIFIER_SHA256SUM='e92a6793c7134c5431c58fbc34700664f101e5c9b1c1fcd93b97978e8b7f88db'
+ENV BUILDOZER_SHA256SUM='3d58a0b6972e4535718cdd6c12778170ea7382de7c75bc3728f5719437ffb84d'
+ENV TENSORFLOW_VERSION='tf-nightly'
+
+# Get the code
+COPY . /tensorboard
```

Comment:
Since you're using `COPY` instead of `git clone` here, you could split up retrieval/installation of your dependencies and pulling in the actual tensorboard application code. This should ensure you don't retrigger the dependency setup steps when you're modifying the rest of the code, which will speed up subsequent builds.

```
# Setup Bazel
COPY ./ci /tensorboard/ci
WORKDIR /tensorboard
RUN ci/download_bazel.sh "${BAZEL_VERSION}" "${BAZEL_SHA256SUM}" ~/bazel
RUN mv ~/bazel /usr/local/bin/bazel && chmod +x /usr/local/bin/bazel && cp ./ci/bazelrc ~/.bazelrc
RUN npm i -g @bazel/ibazel

# Install python dependencies
COPY ./tensorboard/pip_package /tensorboard/tensorboard/pip_package
RUN pip install -r ./tensorboard/pip_package/requirements.txt -r ./tensorboard/pip_package/requirements_dev.txt "$TENSORFLOW_VERSION" && pip freeze --all

# Get the code
COPY . /tensorboard
```

Reply:
This is an interesting idea however installing dependencies then running `COPY . /tensorboard` will not necessarily work if the user has built the project locally and has build artifacts (i.e. `bazel-out`, `node_modules`, etc). I'm going to add a `.dockerignore` as well to try to get the best of both worlds.